cirq-core/cirq/circuits/frozen_circuit.py

```diff
@@ -61,10 +79,32 @@ def _from_moments(cls, moments: Iterable['cirq.Moment']) -> 'FrozenCircuit':
     def moments(self) -> Sequence['cirq.Moment']:
         return self._moments
 
+    @property
+    def tags(self) -> Tuple[Hashable, ...]:
+        """Returns a tuple of the Circuit's tags."""
+        return self._tags
+
+    @property
+    def untagged(self) -> 'cirq.FrozenCircuit':
+        """Returns the underlying operation without any tags."""
+        return self._from_moments(self._moments)
+
+    def with_tags(self, *new_tags: Hashable) -> 'cirq.FrozenCircuit':
+        """Creates a new tagged `FrozenCircuit` with `self.tags` and `new_tags` combined."""
+        if not new_tags:
+            return self
+        new_circuit = FrozenCircuit(tags=self.tags + tuple(new_tags))
+        new_circuit._moments = self._moments
+        return new_circuit
+
     @_compat.cached_method
     def __hash__(self) -> int:
         # Explicitly cached for performance
-        return hash((self.moments,))
+        return hash((self.moments, self.tags)) if self.tags else hash((self.moments,))
```

Comment:
Nit: we can simplify this to `hash((self.moments, self.tags))` there is not going to be noticeable overhead in hashing an empty tuple.

Reply:
done